OCEANWATER-565_Convert_GSAP_CSV_Input_to_Actual_Power_Input_Parameters

jfugate1 opened the pull request:
## Linked Issues:
| EPIC ⚡| [Oceanwater-84 / Power System Modeling](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-84) |
| :---------- | :---------- |
| Jira Ticket 🎟️   | [OCEANWATER-565 / Convert GSAP CSV Input to Actual Power Input Parameters ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-565) |
| Jira Ticket 🎟️   |[OCEANWATER-629 / Address or hide warning messages produced in power_system_node](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-629)
| Github :octocat:  | # |


## Summary of Changes
* Replaced precalculated csv dependency with rostopic subscription to average mechanical power
* Placeholding approximations for voltage and temperature inputs (randomly generated for now)
* Used data to feed into GSAP prognostics algorithm
* Update README.md

## Test
1. Complie, run OW
```bash
roslaunch ow (atacma_y1a.launch | europa_terminator_workspace.launch)
```
2. Subscribe to state_of_charge, remaining_useful_life, and battery_temperature and verify their output
```bash
rostopic echo /power_system_node/state_of_charge
rostopic echo /power_system_node/remaining_useful_life
rostopic echo /power_system_node/battery_temperature
```

Samahu reviewed (dismissed):
It seems that you missed to include the file: `gsap_input_data.csv` as a result I am getting a plenty of "Unable to open data file" messages in the logs when I try to run the simulation.

jfugate1 commented:
> It seems that you missed to include the file: `gsap_input_data.csv` as a result I am getting a plenty of "Unable to open data file" messages in the logs when I try to run the simulation.

Interesting... The "gsap_input_data.csv" file is something that is generated by the code itself. It shouldn't have to be there initially. I will look into this.

jfugate1 commented:
> It seems that you missed to include the file: `gsap_input_data.csv` as a result I am getting a plenty of "Unable to open data file" messages in the logs when I try to run the simulation.

Ahh I see... when I removed all the content from the "data" folder, the folder was removed as well, which is causing the issue. I will fix this.

jfugate1 commented:
@Samahu issue should be resolved with my latest commit

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -13,92 +13,116 @@
+    float max_temp = 21.5; // maximum temp = 21.5 deg. C
+
+    // Temperature estimate based on pseudorandom noise and fixed range
+    float temp_est = min_temp + static_cast <float> (rand()) / (static_cast <float> (RAND_MAX/(max_temp - min_temp)));
```
I don't know where you got this formula to perform range random but it wouldn't give you the desired results
```suggestion
    float temp_est = min_temp + static_cast <float> (rand()) / RAND_MAX * (max_temp - min_temp);
```

jfugate1 replied:
Resolved in latest commit

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -13,92 +13,116 @@
+    bool csv_path_param_exist = nh.param("power_draw_csv_path", csv_path,
+      ros::package::getPath("ow_power_system") + default_csv);
+
+    // Open new file
+    std::ofstream myfile;
+    myfile.open (csv_path);
+    myfile << "Timestamp, power, temperature, voltage\n";
+    // First line is used to initialize GSAP at first time step
+    myfile << update_count << ", " << init_power << ", " << init_temp << ", " << init_voltage << "\n";
+    // Update time step and GSAP initialization parameters
+    update_count = update_count + 1.0;
+    init_power = mech_power;
+    init_temp = temp_est;
+    init_voltage = voltage_est;
+    // Line used to make the prediction (use current time step)
+    myfile << update_count << ", " << mech_power << ", " << temp_est << ", " << voltage_est << "\n";
+    myfile.close();
+
+    //individual msgs to be published
+    std_msgs::Int16 rul_msg;
+    rul_msg.data = 0;  // immediately set to a good default
+    std_msgs::Float64 soc_msg;
+    soc_msg.data = 0.0;
+    std_msgs::Float64 tempbat_msg;
+    tempbat_msg.data = 0.0;
+
+    int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
+
+    // Read battery data from newly created file.
+    auto power_data = read_file(csv_path);
```
I know exactly what you doing here and although it makes good re-use of existing code this is considered lazy programming :facepalm:

Samahu reviewed (changes requested):
You don't need to write the data to a file and reload it just to re-use the **read_file** method.
Simply examine what **read_file** function is doing and fill the same data structure right in memory .. avoiding the write/read to disk which is super slow compared to copying data in memory. This is critically important since this is performed within a callback.

lanseafood commented:
@jfugate1  regarding "Subscribe to power_system_node/state_of_charge, remaining_useful_life, and battery_temperature and verify output "

are there specific values we should be verifying?

jfugate1 commented:
> @jfugate1 regarding "Subscribe to power_system_node/state_of_charge, remaining_useful_life, and battery_temperature and verify output "
> 
> are there specific values we should be verifying?

Good question @lanseafood ! There aren't necessarily specific values to look for since the output now depends on the power output of the system. The values should make sense based on the status of the lander. When actions requiring mechanical power are performed, the RUL and SOC should progressively decrease, whereas they will decrease at a much slower pace when no actions are being performed.

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -14,91 +14,131 @@
-  if (!csv_path_param_exist)
+using namespace chrono;
+
+double update_count = 0.00; // Timestep initialization
+double battery_lifetime = 2738.0; // Estimate of battery lifetime (seconds)
+Datum<double> init_power = 0.00; // Power initialization
+Datum<double> init_temp = 20.00; // Temperature initialization
+Datum<double> init_voltage = 4.10; // Voltage initialization
+auto init_time = system_clock::now(); // Timestep initialization
```
Convert these to member variables of **SubscribeAndPublish** class and initialize them in the constructor

jfugate1 replied:
Resolved in latest commit.

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -14,91 +14,131 @@
+Datum<double> init_voltage = 4.10; // Voltage initialization
+auto init_time = system_clock::now(); // Timestep initialization
+
+class SubscribeAndPublish
```
Use a more descriptive name of the class .. for example **PowerSystemNode**..

jfugate1 replied:
Resolved in latest commit.

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -14,91 +14,131 @@
-    // Validated the parameter
-    if (power_update_rate == 0)
+    using namespace std::chrono;
+    ros::NodeHandle nh;
```
Unreferenced variable
```suggestion
```

jfugate1 replied:
Resolved in latest commit.

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -14,91 +14,131 @@
-    ROS_WARN_NAMED("power_system_node", "Overriding the default power_update_rate!");
-    // Validated the parameter
-    if (power_update_rate == 0)
+    using namespace std::chrono;
```
duplicate statement .. you have already defined this clause early in file
```suggestion
```

jfugate1 replied:
Resolved in latest commit.

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -14,91 +14,131 @@
+    Datum<double> voltage = voltage_est;
+    voltage.setTime(timestamp);
+
+    // Populate initialization data structure
+    initialization_data.insert({MessageId::Watts, init_power});
+    initialization_data.insert({MessageId::Centigrade, init_temp});
+    initialization_data.insert({MessageId::Volts, init_voltage});
+    power_data.push_back(initialization_data);
```
Why do you think that it is necessary to constantly submit the initialization data? I don't understand it .. can you clarify?

Samahu replied:
Just to note that this wasn't the case when we were reading from the csv file .. So I am very skeptic about what is happening in this part of the code.

jfugate1 replied:
Addressed in latest commit.

Samahu commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -14,91 +14,131 @@
+    // Contruct a new prognoser using the prognoser factory. The prognoser
+    // will automatically construct an appropriate model, observer and predictor
+    // based on the values specified in the config.
+    auto prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", config);
```
Why do you need to create and destroy a `prognoser` on each callback?!

jfugate1 replied:
Addressed in latest commit.

Samahu reviewed (changes requested):
- [ ] rename **SubscribeAndPublish** to **PowerSystemNode**
- [ ] Apply [OceanWATERS Code Style](https://babelfish.arc.nasa.gov/confluence/display/OCEANWATERS/Coding+Style+and+Practices)
- [ ] declare prognoser (named m_prognoser) as a member variable for the PowerSystemNode
- [ ] Initialize m_prognoser in the constructor (or anywhere before you subscribe to the mechanical power topic)
- [ ] Re-use m_prognoser within the callback
- [ ] Revise the code with these two points in mind
  - [ ] Is it necessary to submit the initialization data each time?
  - [ ] If it is not necessary then consider dropping the **for** loop

jfugate1 commented on ow_power_system/src/power_system_node.cpp:
```diff
@@ -119,34 +160,45 @@ int main(int argc, char* argv[]) {
+  ros::Publisher SOC_pub;
+  ros::Publisher RUL_pub;
+  ros::Publisher TempBat_pub;
+  ros::Subscriber MechPower_sub;
```
Resolved in latest commit.

jfugate1 commented:
Please note that the latest commit resolves several of the issues described by @Samahu in the previous commit from a stylistic standpoint. However, the functionality needs work. The code compiles, but a crash occurs within the power system node when running due to the way in which the prognoser is initialized.

lanseafood commented:
> Please note that the latest commit resolves several of the issues described by @Samahu in the previous commit from a stylistic standpoint. However, the functionality needs work. The code compiles, but a crash occurs within the power system node when running due to the way in which the prognoser is initialized.

Is that crash something you wanted to resolve in this PR or another?

lanseafood commented:
Ran into these, seems like nothing is being published?

mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/state_of_charge
WARNING: no messages received and simulated time is active.
Is /clock being published?

mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/battery_temperature 
WARNING: no messages received and simulated time is active.
Is /clock being published?

mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/remaining_useful_life 
WARNING: no messages received and simulated time is active.
Is /clock being published?

jfugate1 commented:
> Ran into these, seems like nothing is being published?
> 
> mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/state_of_charge
> WARNING: no messages received and simulated time is active.
> Is /clock being published?
> 
> mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/battery_temperature
> WARNING: no messages received and simulated time is active.
> Is /clock being published?
> 
> mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/remaining_useful_life
> WARNING: no messages received and simulated time is active.
> Is /clock being published?

@lanseafood yes, Ussama was going to add a commit to fix these issues. I was just documenting the current status so he knows. Sorry, should've given you a heads up!

lanseafood commented:
> @lanseafood yes, Ussama was going to add a commit to fix these issues. I was just documenting the current status so he knows. Sorry, should've given you a heads up!

No worries!

Samahu commented:
> > Please note that the latest commit resolves several of the issues described by @Samahu in the previous commit from a stylistic standpoint. However, the functionality needs work. The code compiles, but a crash occurs within the power system node when running due to the way in which the prognoser is initialized.
> 
> Is that crash something you wanted to resolve in this PR or another?

I'll try to address them as part of this PR .. it maybe till Thursday until I submit a new request for review.

Samahu commented:
To reviewers: I have completed the described work for this ticket and the power_system_node should run live. I did refactor parts of the code while doing so.
Additionally, I have added the 3 following tickets to the current sprint:
- [OCEANWATER-651 / Read initial battery voltage and battery temperature from config](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-651)
- [OCEANWATER-652 / Remove the strong dependency of PowerSystemNode on the mechanical power topic](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-652)
- [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653)

@jfugate1 I can not add you as a reviewer of this PR since you created the pull request but you should be able to give feedback as any other reviewer. I have also added Mike

jfugate1 commented:
> To reviewers: I have completed the described work for this ticket and the power_system_node should run live. I did refactor parts of the code while doing so.
> Additionally, I have added the 3 following tickets to the current sprint:
> 
> * [OCEANWATER-651 / Read initial battery voltage and battery temperature from config](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-651)
> * [OCEANWATER-652 / Remove the strong dependency of PowerSystemNode on the mechanical power topic](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-652)
> * [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653)
> 
> @jfugate1 I can not add you as a reviewer of this PR since you created the pull request but you should be able to give feedback as any other reviewer. I have also added Mike

Thanks, Ussama! I will get this review done as soon as possible and comment accordingly. I also figured it would make sense for you to eliminate the stylistic errors that produce warnings when building the power system node (if you haven't already done so).

Samahu commented:
> > To reviewers: I have completed the described work for this ticket and the power_system_node should run live. I did refactor parts of the code while doing so.
> > Additionally, I have added the 3 following tickets to the current sprint:
> > 
> > * [OCEANWATER-651 / Read initial battery voltage and battery temperature from config](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-651)
> > * [OCEANWATER-652 / Remove the strong dependency of PowerSystemNode on the mechanical power topic](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-652)
> > * [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653)
> > 
> > @jfugate1 I can not add you as a reviewer of this PR since you created the pull request but you should be able to give feedback as any other reviewer. I have also added Mike
> 
> Thanks, Ussama! I will get this review done as soon as possible and comment accordingly. I also figured it would make sense for you to eliminate the stylistic errors that produce warnings when building the power system node (if you haven't already done so).

I haven't yet but I'll get to that

jfugate1 commented:
Verified the power system node now works as advertised. Once we can get the style casting warnings removed, I believe this will be ready to merge.

Samahu commented:
> Verified the power system node now works as advertised. Once we can get the style casting warnings removed, I believe this will be ready to merge.

I have resolved [OCEANWATER-629 / Address or hide warning messages produced in power_system_node](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-629).
All of these warnings except one are outside our code. I addressed the one that was in our code and for the rest I disable the cmake flag causing this issue.

jfugate1 commented:
> > Verified the power system node now works as advertised. Once we can get the style casting warnings removed, I believe this will be ready to merge.
> 
> I have resolved [OCEANWATER-629 / Address or hide warning messages produced in power_system_node](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-629).
> All of these warnings except one are outside our code. I addressed the one that was in our code and for the rest I disable the cmake flag causing this issue.

Nice! Confirmed. Consider the PR approved on my end.

kmdalal commented on ow_power_system/include/power_system_node.h:
```diff
@@ -0,0 +1,46 @@
+class PowerSystemNode
+{
+public:
+  PowerSystemNode();
```
It's good to also be explicit about the copy constructor, assignment operator, and destructor.  I.e. define them or 'delete' them or a comment that you want to use the defaults.

Samahu replied:
Yes, originally there was some initialization in the constructor. But I have moved things around and the constructor is no longer utilized .. I will omit it!

Samahu replied:
addressed

kmdalal commented on ow_power_system/include/power_system_node.h:
```diff
@@ -0,0 +1,46 @@
+private:
+  void powerCallback(const std_msgs::Float64::ConstPtr& msg);
+
+private:
```
A comment I've made before, but it's odd to see repeated "private:" etc.  Not idiomatic C++.

Samahu replied:
Removed the extra private

kmdalal reviewed (approved):
Made some minor code comments.  The power node worked as described, and the published values looked reasonable.  Great to see this in the system!
